### PR TITLE
Martin/ios 400 after logging in we are still connected

### PR DESCRIPTION
**Pull request by MartinJibber**
This fixes an issue where the anonymous user was still logged in after finishing onboarding.

It also makes the main coordinator more robust so that waitlist user's are prevented from getting to the whole app.

### CODE REVIEW

File: Shared/Coordinator/MainCoordinator+Launch.swift [hunk 1]
```diff
@@ -8,85 +8,28 @@
 
 import Foundation
 import StreamChat
-import Intents
 
 #if IOS
 extension MainCoordinator {
 
-    func handle(result: LaunchStatus) {
```
Comment (MartinJibber) [inline Shared/Coordinator/MainCoordinator+Launch.swift:16]: I moved this to the logic to the main file.

File: Shared/Coordinator/MainCoordinator+Launch.swift [hunk 2]
```diff
@@ -8,85 +8,28 @@
...
-                                                              startingMessageId: startingMessageId)
-                self.addChildAndStart(coordinator, finishedHandler: { (_) in })
-
-                self.router.setRootModule(coordinator)
-
-                await self.checkForPermissions()
-            }
-        } else {
+        // Ensure that the chat client is initialized for the logged in user.
+        if !ChatClient.isConnected || ChatClient.shared.currentUserId != User.current()?.userObjectId {
```
Comment (MartinJibber) [inline Shared/Coordinator/MainCoordinator+Launch.swift:18]: Fixes the anonymous user issue.

File: Shared/Coordinator/MainCoordinator.swift [hunk 3]
```diff
@@ -12,8 +12,6 @@ import Parse
 
 class MainCoordinator: Coordinator<Void> {
 
-    lazy var splashVC = SplashViewController()
```
Comment (MartinJibber) [inline Shared/Coordinator/MainCoordinator.swift:15]: Moved this to a dedicated launch coordinator.

File: Shared/Coordinator/MainCoordinator.swift [hunk 4]
```diff
@@ -24,93 +22,113 @@ class MainCoordinator: Coordinator<Void> {
...
+        } else if user.status == .waitlist {
+            // The user finished onboarding but is on the waitlist so don't proceed to the main app.
+            self.runWaitlistFlow()
+            return
+        }
 
+        // As a final catch-all, make sure the user is fully activated.
+        guard user.status == .active else {
+            self.runOnboardingFlow()
+            return
```
Comment (MartinJibber) [inline Shared/Coordinator/MainCoordinator.swift:86]: Before we deep link into a target, we need to get past all of these checks.

File: iOS/_Stream Temp/ChatClient+Extensions.swift [hunk 5]
```diff
@@ -117,21 +115,6 @@ extension ChatClient {
...
-                if let e = error {
-                    continuation.resume(throwing: e)
-                } else {
-                    continuation.resume(returning: ())
-                }
-            }
-        })
-    }
-
     func messageController(for messageable: Messageable) -> MessageController? {
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ChatClient+Extensions.swift:118]: Unused.

File: iOS/Flows/Conversation/ConversationListCoordinator.swift [hunk 6]
```diff
@@ -266,3 +271,26 @@ private class ModalPhotoViewController: PhotoViewController {
...
+
+    @MainActor
+    private func presentPermissions() {
+        let coordinator = PermissionsCoordinator(router: self.router, deepLink: self.deepLink)
+        self.addChildAndStart(coordinator) { [unowned self] result in
+            self.router.dismiss(source: self.conversationListVC, animated: true)
+        }
+        self.router.present(coordinator, source: self.conversationListVC)
+    }
+}
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationListCoordinator.swift:296]: Moved this into the conversation list coordinator.

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**